examples/go/vm-reconfig-via-tag/handler/handler.go

```diff
@@ -0,0 +1,564 @@
+// a VM config containing the config type and value.
+func buildConfig(attachedTag tags.Tag, cfgType string) (vmConfig, error) {
+	if cfgType == "numCPU" || cfgType == "memoryMB" || cfgType == "numCoresPerSocket" {
+		val, err := strconv.Atoi(attachedTag.Name)
+		if err != nil {
+			return vmConfig{}, fmt.Errorf("convert string to int :%w", err)
+		}
+
+		return vmConfig{name: cfgType, valInt: val}, nil
+	}
+
+	if cfgType == "memoryHotAddEnabled" || cfgType == "cpuHotRemoveEnabled" || cfgType == "cpuHotAddEnabled" {
+		val, err := strconv.ParseBool(attachedTag.Name)
+		if err != nil {
+			return vmConfig{}, fmt.Errorf("convert string to bool: %w", err)
+		}
+
+		return vmConfig{name: cfgType, valBool: val}, nil
+	}
+
+	return vmConfig{}, errors.New(cfgType + " is not an expected config type")
+}
+
+// filterOutCurrentConfigs removes the tag configs that are already current configs.
+func filterOutCurrentConfigs(tagConfigs, currConfigs []vmConfig) []vmConfig {
+	unappliedCfgs := []vmConfig{}
+
+	for _, tc := range tagConfigs {
+		if !isCfgCurrent(tc, currConfigs) {
+			unappliedCfgs = append(unappliedCfgs, tc)
+		}
+	}
+	return unappliedCfgs
+}
+
+// isHwCfgMatch determines if the given tag's hardware configuration
+// is already the current configuration of the hardware.
+func isCfgCurrent(tagConfig vmConfig, currConfigs []vmConfig) bool {
+	for _, curr := range currConfigs {
+		if tagConfig == curr {
+			return true
+		}
+	}
+
+	return false
+}
+
+// makeCfgsMatch sets configuration of the VM to that of the attached tag.
+func (c *vsClient) applyCfgs(ctx context.Context, moVM mo.VirtualMachine, mor types.ManagedObjectReference, cfgs []vmConfig) (*object.Task, error) {
+	vm := object.NewVirtualMachine(c.govmomi.Client, mor)
+	desiredSpec := generateDesiredSpec(cfgs, moVM)
+
+	task, err := vm.Reconfigure(ctx, desiredSpec)
+	if err != nil {
+		return nil, err
+	}
+
+	if debug {
+		log.Println("VM reconfigured with", desiredSpec)
+	}
+
+	return task, nil
+}
+
+func generateDesiredSpec(cfgs []vmConfig, moVM mo.VirtualMachine) types.VirtualMachineConfigSpec {
+	var spec types.VirtualMachineConfigSpec
+
+	for _, c := range cfgs {
+		switch c.name {
+		case "numCPU":
+			spec.NumCPUs = int32(c.valInt)
+		case "memoryMB":
+			spec.MemoryMB = int64(c.valInt)
+		case "numCoresPerSocket":
+			spec.NumCoresPerSocket = int32(c.valInt)
+		case "memoryHotAddEnabled":
+			spec.MemoryHotAddEnabled = &c.valBool
+		case "cpuHotRemoveEnabled":
+			spec.CpuHotRemoveEnabled = &c.valBool
+		case "cpuHotAddEnabled":
+			spec.CpuHotAddEnabled = &c.valBool
+		}
+	}
+
+	// Set ChangeVersion to guard against updates that have happened between when
+	// configInfo is read and when it is applied.
+	spec.ChangeVersion = moVM.Config.ChangeVersion
+
+	return spec
+}
+
+func appliedConfigsMessage(task *object.Task, cfgs []vmConfig) string {
+	if task == nil {
+		return "Nothing to configure."
+	}
+
+	msg := task.String() + " set "
```

Comment:
Nit: (structured) JSON string response would be nicer.

Reply:
Yes, I agree. I'll marshal the message to be in a JSON string.